src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,351 @@
+        .into_iter()
+        .chain(
+            try_join_all(input.iter().skip(1).enumerate().map(|(i, x)| {
+                let c = stop_bit_context.clone();
+                let record_id = RecordId::from(i);
+                async move {
+                    c.multiply(record_id, &x.is_trigger_report, &x.helper_bit)
+                        .await
+                }
+            }))
+            .await?,
+        )
+        .collect::<Vec<_>>();
+
+    // First, create a vector of timedeltas. This vector contains non-zero values only for
+    // rows with `stop_bit` = 1, meaning that the row is a trigger event, and has the same
+    // match key as the event one above.
+    let t_delta_context = ctx
+        .narrow(&Step::InitializeTimeDelta)
+        .set_total_records(num_rows - 1);
+    let mut t_delta = Some(T::ZERO)
+        .into_iter()
+        .chain(
+            try_join_all(
+                zip(input.iter(), input.iter().skip(1))
+                    .zip(stop_bits.iter().skip(1))
+                    .enumerate()
+                    .map(|(i, ((prev, curr), b))| {
+                        let c = t_delta_context.clone();
+                        let record_id = RecordId::from(i);
+                        let delta = curr.timestamp.clone() - &prev.timestamp;
+                        async move { c.multiply(record_id, &delta, b).await }
+                    }),
+            )
+            .await?,
+        )
+        .collect::<Vec<_>>();
+
+    // Do the prefix-sum accumulation of timedeltas. This is similar to
+    // `do_the_binary_tree_thing()`, but it sums up values in top-down order.
+    for (depth, step_size) in std::iter::successors(Some(1_usize), |prev| prev.checked_mul(2))
+        .take_while(|&v| v < num_rows)
+        .enumerate()
+    {
+        let record_count = num_rows - step_size;
+
+        let depth_i_ctx = ctx
+            .narrow(&InteractionPatternStep::from(depth))
+            .set_total_records(record_count);
+        let mt_ctx = depth_i_ctx.narrow(&Step::PreviousStopBitTimesCurrent);
+
+        let mut futures = Vec::with_capacity(record_count);
+
+        // `i` starts from 0 for RecordId. The actual logic skips the first
+        // `step_size` rows.
+        for i in 0..record_count {
+            let index = i + step_size;
+
+            let c1 = depth_i_ctx.clone();
+            let c2 = mt_ctx.clone();
+            let record_id = RecordId::from(i);
+
+            let previous_stop_bit = &stop_bits[index - step_size];
+            let current_stop_bit = &stop_bits[index];
+            let previous_t_delta = &t_delta[index - step_size];
+
+            futures.push(async move {
+                try_join(
+                    // if the current stop_bit is 0, don't accumulate
+                    c1.multiply(record_id, previous_t_delta, current_stop_bit),
+                    // next `stop_bit`
+                    c2.multiply(record_id, previous_stop_bit, current_stop_bit),
```

Comment:
We definitely need to call the same `do_the_binary_tree_thing` function. This is missing so many of the optimizations we put into that thing, like the whole "don't compute the stop bits on the final iteration" stuff.

Reply:
Oh, and also the whole "don't compute stop bits on the final record" optimization.